pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTestNew.java

```diff
@@ -0,0 +1,1543 @@
+    nPartitions = 4;
+    instances = getInstanceList(3);
+    exceptionExpected = true;
+    validateLLCNewTable(tableConfig, idealState, nPartitions, nReplicas, instances, exceptionExpected,
+        skipNewPartitions);
+
+    // noop path - 0 partitions
+    tableConfig = makeTableConfig(tableName, nReplicas, KAFKA_OFFSET, DUMMY_HOST, DEFAULT_SERVER_TENANT,
+        DEFAULT_STREAM_ASSIGNMENT_STRATEGY);
+    idealState = idealStateBuilder.build();
+    nPartitions = 0;
+    exceptionExpected = false;
+    validateLLCNewTable(tableConfig, idealState, nPartitions, nReplicas, instances, exceptionExpected,
+        skipNewPartitions);
+
+    // noop path - ideal state disabled
+    tableConfig = makeTableConfig(tableName, nReplicas, KAFKA_OFFSET, DUMMY_HOST, DEFAULT_SERVER_TENANT,
+        DEFAULT_STREAM_ASSIGNMENT_STRATEGY);
+    idealState = idealStateBuilder.disableIdealState().build();
+    nPartitions = 4;
+    exceptionExpected = false;
+    validateLLCNewTable(tableConfig, idealState, nPartitions, nReplicas, instances, exceptionExpected,
+        skipNewPartitions);
+
+    // happy paths - new table config with nPartitions and sufficient instances
+    tableConfig = makeTableConfig(tableName, nReplicas, KAFKA_OFFSET, DUMMY_HOST, DEFAULT_SERVER_TENANT,
+        DEFAULT_STREAM_ASSIGNMENT_STRATEGY);
+    idealState = idealStateBuilder.build();
+    nPartitions = 4;
+    exceptionExpected = false;
+    validateLLCNewTable(tableConfig, idealState, nPartitions, nReplicas, instances, exceptionExpected,
+        skipNewPartitions);
+
+    idealState = idealStateBuilder.build();
+    nPartitions = 8;
+    validateLLCNewTable(tableConfig, idealState, nPartitions, nReplicas, instances, exceptionExpected,
+        skipNewPartitions);
+
+    idealState = idealStateBuilder.build();
+    nPartitions = 8;
+    instances = getInstanceList(10);
+    validateLLCNewTable(tableConfig, idealState, nPartitions, nReplicas, instances, exceptionExpected,
+        skipNewPartitions);
+
+    idealState = idealStateBuilder.build();
+    nPartitions = 12;
+    validateLLCNewTable(tableConfig, idealState, nPartitions, nReplicas, instances, exceptionExpected,
+        skipNewPartitions);
+  }
+
+  private IdealState validateLLCNewTable(TableConfig tableConfig, IdealState idealState, int nPartitions, int nReplicas,
+      List<String> instances, boolean exceptionExpected, boolean skipNewPartitions) {
+    IdealState updatedIdealState = null;
+    FakePinotLLCRealtimeSegmentManager segmentManager = new FakePinotLLCRealtimeSegmentManager(null);
+    segmentManager._partitionAssignmentGenerator.setConsumingInstances(instances);
+    try {
+      updatedIdealState = segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+      Assert.assertFalse(exceptionExpected);
+      Map<String, LLCSegmentName> partitionToLatestSegments =
+          segmentManager._partitionAssignmentGenerator.getPartitionToLatestSegments(updatedIdealState);
```

Comment:
As of now the partitionAssignmentGenerator is a field inside the PinotLLCRealtimeSegmentManager, and hence accessed in this way. When we move to having a factory for partition assignment generators, this will be refactored accordingly